string k_name
---
string[] keys
